### ISSUE TITLE
[Feature]: Allow head_size smaller than 128 on TPU with Pallas backend

### ISSUE THREAD

**Issue by manninglucas** [vllm-project/vllm#10343]
### 🚀 The feature, motivation and pitch

I would like to serve smaller models (e.g facebook/opt-125m) using VLLM on TPU. I can't do this currently because the Pallas backend has the limitation `NotImplementedError: Head size must be a multiple of 128`. I can't find a reason why this limitation is in place, and it would be great to be able to remove it with a flag or entirely. If my understanding is incorrect and there is a reason to have this limitation in place, please let me know! Thanks for your work on VLLM.

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by manninglucas** [comment]
Here is a code pointer to the limitation FWIW.

https://github.com/vllm-project/vllm/blob/main/vllm/attention/backends/pallas.py#L112

**Comment by bvrockwell** [comment]
Sorry @manninglucas for missing this! I had been filtering on the TPU tag. Which model are you trying to serve? I don't think we've encountered this because the head dim is usually 128.

**Comment by manninglucas** [comment]
Here is a non-comprehensive list of models with head dim < 128:
- [LLama-3.2-1B](https://huggingface.co/meta-llama/Llama-3.2-1B/blob/main/config.json)
- [Qwen2.5-0.5B](https://huggingface.co/Qwen/Qwen2.5-0.5B/blob/main/config.json)
- [MobileLLM-1B](https://huggingface.co/facebook/MobileLLM-1B/blob/main/config.json)
- [Phi 2](https://huggingface.co/microsoft/phi-2/blob/main/config.json)

**Comment by miladm** [comment]
cc @vanbasten23 to help take a look at this issue.

**Comment by ayushgun** [comment]
> Here is a non-comprehensive list of models with head dim < 128:
> 
>     * [LLama-3.2-1B](https://huggingface.co/meta-llama/Llama-3.2-1B/blob/main/config.json)
> 
>     * [Qwen2.5-0.5B](https://huggingface.co/Qwen/Qwen2.5-0.5B/blob/main/config.json)
> 
>     * [MobileLLM-1B](https://huggingface.co/facebook/MobileLLM-1B/blob/main/config.json)
> 
>     * [Phi 2](https://huggingface.co/microsoft/phi-2/blob/main/config.json)

"llava-hf/llava-onevision-qwen2-0.5b-ov-h" as well too.

**Comment by vanbasten23** [comment]
I see. In your example, the "head_dim" is 64. The reason for this constraints "Head size must be a multiple of 128" is due to a constraint of Pallas kernel, namely
```
On TPU, only blocks with rank at least 1 are supported. Furthermore, the last two dimensions of your block shape must be equal to the respective dimension of the overall array, or be divisible by 8 and 128 respectively.
```
per https://jax.readthedocs.io/en/latest/pallas/grid_blockspec.html.
In the [paged attention kernel](https://github.com/jax-ml/jax/blob/main/jax/experimental/pallas/ops/tpu/paged_attention/paged_attention_kernel.py), the head_dim happens to be the last dimension which has to be a multiple of 128.

To workaround, you can pad the head_size. But for long term solution, I think the kernel could be updated so that it pad internally the head_dim to be the closest multiple of 128.

**Comment by ewianda** [comment]
@vanbasten23 will be kind enough to demonstrate how the head_size can be padded.

**Comment by adityarajsahu** [comment]
Hi @vanbasten23, can you please provide any code snippets on how to pad the head?

**Comment by bvrockwell** [comment]
this should be available very soon cc @bythew3i @yaochengji

**Comment by yaochengji** [comment]
@bvrockwell it's already supported. E.g. phi-3 has head_dim = 96.

**Comment by bythew3i** [comment]
> @bvrockwell it's already supported. E.g. phi-3 has head_dim = 96.

Yes, any head_dim <= 128 should already supported! @yaochengji added this feature weeks ago.

**Comment by bvrockwell** [comment]
> > [@bvrockwell](https://github.com/bvrockwell) it's already supported. E.g. phi-3 has head_dim = 96.
> 
> Yes, any head_dim <= 128 should already supported! [@yaochengji](https://github.com/yaochengji) added this feature weeks ago.

oops, my bad! This is great 👍 thanks @bythew3i  I suppose we can close this issue if so @yaochengji ?

**Comment by yaochengji** [comment]
Yeah, we can close it because it's already supported.
